File: common/src/main/java/com/tianshaokai/common/utils/SDUtils.java
package com.tianshaokai.common.utils;

import android.content.Context;
import android.os.Environment;
import android.os.StatFs;

import java.io.File;

public class SDUtils {

    public static boolean isSdMounted() {
        return Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState());
    }

    /**
     * @return 获取外部相册路径
     */
    public static String getStorageDCIMPath() {
        return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM).getAbsolutePath();
    }

    /**
     * @return 获取外部音频路径
     */
    public static String getStorageAudioPath() {
        return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC).getAbsolutePath();
    }

    /**
     * @return 获取外部视频路径
     */
    public static String getStorageVideoPath() {
        return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MOVIES).getAbsolutePath();
    }


    public static String getExternalFilesDir(Context context, String dir) {
        if (context == null) {
            return "";
        }
        File filePath = context.getApplicationContext().getExternalFilesDir(dir);
        if (filePath != null) {
            return filePath.getAbsolutePath();
        }
        return "";
    }

    public static String getPackageDCIMPath(Context context) {
        if (context == null) {
            return "";
        }
        File filePath = context.getApplicationContext().getExternalFilesDir(Environment.DIRECTORY_DCIM);
        if (filePath != null) {
            return filePath.getAbsolutePath();
        }
        return "";
    }

    public static String getPackageAudioPath(Context context) {
        if (context == null || !isSdMounted()) {
            return "";
        }
        File filePath = context.getApplicationContext().getExternalFilesDir(Environment.DIRECTORY_MUSIC);
        if (filePath != null) {
            return filePath.getAbsolutePath();
        }
        return "";
    }

    public static String getPackageMoviePath(Context context) {
        if (context == null) {
            return "";
        }
        File filePath = context.getApplicationContext().getExternalFilesDir(Environment.DIRECTORY_MOVIES);
        if (filePath != null) {
            return filePath.getAbsolutePath();
        }
        return "";
    }

    public static String getPackageCrashPath(Context context) {
        return getExternalFilesDir(context, "Crash");
    }

    /**
     * 未使用磁盘空间
     * @return 返回剩余空间
     */
    public static long getFreeDiskSpace() {
        if (isSdMounted()) {
            File sdcardDir = Environment.getExternalStorageDirectory();
            StatFs sf = new StatFs(sdcardDir.getPath());
            long blockSize = sf.getBlockSize();
            long availCount = sf.getAvailableBlocks();
            return availCount * blockSize;
        }
        return 0;
    }

}

File: framework/src/main/java/com/tianshaokai/framework/util/SDUtils.kt
package com.tianshaokai.framework.util

import android.content.Context
import android.os.Environment
import android.os.StatFs

object SDUtils {

    fun isSdMounted(): Boolean {
        return Environment.MEDIA_MOUNTED == Environment.getExternalStorageState()
    }

    fun getExternalFilesDir(context: Context, dir: String): String {
        if (!isSdMounted()) {
            return ""
        }
        val filePath = context.applicationContext.getExternalFilesDir(dir)
        return if (filePath != null) {
            filePath.absolutePath
        } else ""
    }

    fun getPackageDCIMPath(context: Context): String {
       return getExternalFilesDir(context, Environment.DIRECTORY_DCIM)
    }

    fun getPackageAudioPath(context: Context): String {
        return getExternalFilesDir(context, Environment.DIRECTORY_MUSIC)
    }

    fun getPackageMoviePath(context: Context): String {
        return getExternalFilesDir(context, Environment.DIRECTORY_MOVIES)
    }

    fun getPackageCrashPath(context: Context): String {
        return getExternalFilesDir(context, "Crash")
    }

    /**
     * 获取剩余存储，有外部取外部，没有取内部
     * 返回 字节 Bytes
     */
    fun getMemorySize(): Long {
        return if (isSdMounted()) {
            getAvailableExternalMemorySize()
        } else {
            getAvailableInternalMemorySize()
        }
    }

    /**
     * 获取sdcard剩余存储空间
     */
    fun getAvailableExternalMemorySize(): Long {
        return if (isSdMounted()) {
            try {
                val path = Environment.getExternalStorageDirectory()
                val stat = StatFs(path.path)
                val blockSize = stat.blockSizeLong
                val availableBlocks = stat.availableBlocksLong
                availableBlocks * blockSize
            } catch (e: Exception) {
                0L
            }
        } else {
            0L
        }
    }

    /**
     * 获取手机内部剩余存储空间
     */
    fun getAvailableInternalMemorySize(): Long {
        val path = Environment.getDataDirectory()
        val stat = StatFs(path.path)
        val blockSize = stat.blockSizeLong
        val availableBlocks = stat.availableBlocksLong
        return availableBlocks * blockSize
    }

    /**
     * 获取手机内部总的存储空间
     */
    fun getTotalInternalMemorySize(): Long {
        val path = Environment.getDataDirectory()
        val stat = StatFs(path.path)
        val blockSize = stat.blockSizeLong
        val totalBlocks = stat.blockCountLong
        return totalBlocks * blockSize
    }
}